### PR TITLE
docs: add more detail to the CONTRIBUTING docs

### PR DESCRIPTION
**Pull request by petebacondarwin**
Also ensure that the package.json scripts are consistent.

---

By the way, I looked into blocking the use of `npm install` directly in packages and could find no nice way to do it. Any `install` or `preinstall` scripts in package.json do not run until after the node_modules directory has been filled; and also these scripts will run even if you do the `npm install` from the top level, so you would need a way to flag to the script not to run in this case. And moreover, these scripts would be included in the distributable of wrangler, which would mean they would run when you installed wrangler as a user.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,12 +1,204 @@
...
+Wrangler2 is an open source project and we welcome contributions from you. Thank you!
 
-# Steps For Making Changes
+Below you can find some guidance on how to be most effective when contributing to the project.
+
+## Getting started
+
+### Set up your environment
+
+Wrangler2 is built and run on the node.js JavaScript runtime.
```
Comment (GregBrimble) [inline CONTRIBUTING.md:11]: Nit: `Node.js` is capitalized (not `node.js`).

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,12 +1,204 @@
...
+
+- Install the latest LTS version of [node.js](https://nodejs.dev/) - we recommend using a Node version manager like [nvm](https://github.com/nvm-sh/nvm).
+- Install a code editor - we recommend using [VS Code](https://code.visualstudio.com/).
+- Install the [git](https://git-scm.com/) version control tool.
+
+### Fork and clone this repository
+
+Any contributions you make will be via [Pull Requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) on [Github](https://github.com/) developed in a local git repository and pushed to your own fork of the repository.
```
Comment (GregBrimble) [inline CONTRIBUTING.md:19]:
```suggestion
Any contributions you make will be via [Pull Requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) on [GitHub](https://github.com/) developed in a local git repository and pushed to your own fork of the repository.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,12 +1,204 @@
...
+
+- Ensure you have [created an account](https://docs.github.com/en/get-started/onboarding/getting-started-with-your-github-account) on Github.
```
Comment (GregBrimble) [inline CONTRIBUTING.md:21]:
```suggestion
- Ensure you have [created an account](https://docs.github.com/en/get-started/onboarding/getting-started-with-your-github-account) on GitHub.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,12 +1,204 @@
...
-# Steps For Making Changes
+Below you can find some guidance on how to be most effective when contributing to the project.
+
+## Getting started
+
+### Set up your environment
+
+Wrangler2 is built and run on the node.js JavaScript runtime.
+
+- Install the latest LTS version of [node.js](https://nodejs.dev/) - we recommend using a Node version manager like [nvm](https://github.com/nvm-sh/nvm).
```
Comment (GregBrimble) [inline CONTRIBUTING.md:13]:
```suggestion
- Install the latest LTS version of [Node.js](https://nodejs.dev/) - we recommend using a Node version manager like [nvm](https://github.com/nvm-sh/nvm).
```

File: CONTRIBUTING.md (hunk 1 above)
Comment (GregBrimble) [reply CONTRIBUTING.md:11]:
```suggestion
Wrangler2 is built and run on the Node.js JavaScript runtime.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,12 +1,204 @@
...
+
+- Type check all the code in the repository
+  ```sh
+  > npm run type-check
+  ```
+- VS Code will also run type-checking while editing source code, providing immediate feedback.
+
+### Linting
+
+The code is checked for linting errors by [eslint](https://eslint.org/).
```
Comment (GregBrimble) [inline CONTRIBUTING.md:105]:
```suggestion
The code is checked for linting errors by [ESLint](https://eslint.org/).
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,12 +1,204 @@
...
+
+- Run the linting checks
+  ```sh
+  > npm run lint
+  ```
+- The repository has a recommended VS Code plugin to run eslint checks while editing source code, providing immediate feedback.
+
+### Formatting
+
+The code is checked for formatting errors by [prettier](https://prettier.io/).
```
Comment (GregBrimble) [inline CONTRIBUTING.md:115]:
```suggestion
The code is checked for formatting errors by [Prettier](https://prettier.io/).
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,12 +1,204 @@
...
+
+- Run the formatting checks
+  ```sh
+  > npm run format-check
+  ```
+- The repository has a recommended VS Code plugin to run prettier checks, and to automatically format using prettier, while editing source code, providing immediate feedback.
```
Comment (GregBrimble) [inline CONTRIBUTING.md:121]:
```suggestion
- The repository has a recommended VS Code plugin to run Prettier checks, and to automatically format using Prettier, while editing source code, providing immediate feedback.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,12 +1,204 @@
...
+
+### Testing
+
+Tests in a workspace are executed, by [jest](https://jestjs.io/), which is configured to automatically compile and bundle the TypeScript before running the tests.
```
Comment (GregBrimble) [inline CONTRIBUTING.md:125]:
```suggestion
Tests in a workspace are executed, by [Jest](https://jestjs.io/), which is configured to automatically compile and bundle the TypeScript before running the tests.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,12 +1,204 @@
...
+  - Or add and commit files via the command line
+  ```sh
+  > git add <paths-to-changes-files>
+  > git commit
+  ```
+- Push changes to your fork
+  ```sh
+  git push -u origin <new-branch-name>
+  ```
+- Once you are happy with your changes, create a Pull Request on Github
```
Comment (GregBrimble) [inline CONTRIBUTING.md:165]:
```suggestion
- Once you are happy with your changes, create a Pull Request on GitHub.
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,12 +1,204 @@
...
+  ```
+- Watch the files in a specific workspace (e.g. wrangler), and run the tests when anything changes
+  ```sh
+  > npm run test -w wrangler -- --watch
+  ```
+
+## Steps For Making Changes
+
+Every change you make should be stored in a [git commit](https://github.com/git-guides/git-commit).
+Changes should be committed to a new local branch, which then gets pushed to your fork of the repository on Github.
```
Comment (GregBrimble) [inline CONTRIBUTING.md:143]:
```suggestion
Changes should be committed to a new local branch, which then gets pushed to your fork of the repository on GitHub.
```

**Comment by GregBrimble** [comment]
Looks great! Only thing I'd consider adding would be instructions (or a link to [these instructions](https://code.visualstudio.com/docs/editor/extension-marketplace#_workspace-recommended-extensions)) on how to install recommended extensions.

File: CONTRIBUTING.md (hunk 1 above)
Comment (petebacondarwin) [reply CONTRIBUTING.md:11]: Right! Thanks.

File: CONTRIBUTING.md (hunk 2 above)
Comment (petebacondarwin) [reply CONTRIBUTING.md:19]: 👍

File: CONTRIBUTING.md (hunk 5 above)
Comment (petebacondarwin) [reply CONTRIBUTING.md:105]: 😆

File: CONTRIBUTING.md (hunk 6 above)
Comment (petebacondarwin) [reply CONTRIBUTING.md:115]: 🤣

**Comment by petebacondarwin** [comment]
Thanks for the feedback @GregBrimble !

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,12 +1,205 @@
...
+
+When doing normal development you may want to run these checks individually.
+
+### Type Checking
+
+The code is checked for type errors by [TypeScript](https://www.typescriptlang.org/).
+
+- Type check all the code in the repository
+  ```sh
+  > npm run type-check
```
Comment (threepointone) [inline CONTRIBUTING.md:100]: This doesn't exist. Should we replace with `npx tsc`, or actually make a `type-check` script? Either works for me.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,12 +1,205 @@
...
+  ```
+- The repository has a recommended VS Code plugin to run ESLint checks while editing source code, providing immediate feedback.
+
+### Formatting
+
+The code is checked for formatting errors by [Prettier](https://prettier.io/).
+
+- Run the formatting checks
+  ```sh
+  > npm run format-check
```
Comment (threepointone) [inline CONTRIBUTING.md:120]: Like above, should we make an npm script for this?

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,12 +1,205 @@
...
+  ```
+- The repository has a recommended VS Code plugin to run Prettier checks, and to automatically format using Prettier, while editing source code, providing immediate feedback.
+
```
Comment (threepointone) [inline CONTRIBUTING.md:123]: For the above 3, I wonder if we should have this much detail at all. We want to encourage IDE driven development (I assume?) and the cli driven commands aren't as accessible. I dunno... I don't feel strongly about it, but still.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,12 +1,205 @@
...
+We use the [`changesets`](https://github.com/changesets/changesets/blob/main/README.md) tool for creating changesets, publishing versions and updating the changelog.
+
+- Create a changeset for the current change.
+  ```sh
+  > npm run changeset
+  ```
+  or
+  ```sh
+  > npx changeset
+  ```
```
Comment (threepointone) [inline CONTRIBUTING.md:180]: Let's just recommend one of these.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,12 +1,205 @@
...
+
+[BREAKING CHANGES <BREAKING_CHANGE_NOTES>]
+```
 
-- Changeset Bot
-- Create a changeset by running `npm run changeset` or `npx changeset`. [More info.](https://github.com/changesets/changesets#how-do-i-get-started)
+- `TYPE` should be a single word describing the "type" of the change. For example, one of `feature`, `fix`, `refactor`, `docs` or `chore`.
+- `TITLE` should be a single sentence containing an imperative description of the change.
+- `BODY` should be one or more paragraphs that go into more detail about the reason for the change and anything notable about the approach taken.
+- `BREAKING_CHANGE_NOTES` (optional) should be one or more paragraphs describing how this change breaks current usage and how to migrate to the new usage.
```
Comment (threepointone) [inline CONTRIBUTING.md:205]: This is a good set of rules. I'll start following them myself 😅

File: package.json [hunk 16]
```diff
@@ -29,11 +29,14 @@
...
+    "type-check": "tsc",
     "lint": "eslint packages/**",
-    "check": "prettier packages/** --check && tsc && npm run lint && npm run build && npm run test",
+    "format-check": "prettier packages/** --check",
+    "check": "npm run format-check && npm run type-check && npm run lint && npm run build && npm run test",
     "prettify": "prettier packages/** --write",
     "build": "npm run build --workspace=wrangler",
-    "test": "npm run test --workspace=wrangler"
+    "test": "npm run test --workspaces --if-present",
+    "changeset": "changeset"
```
Comment (threepointone) [inline package.json:39]: oh you did actually add the scripts here.

**Review by threepointone — APPROVED** [review]
Stamping so we can land and iterate on this, this is a great step forward. Thank you very much!

File: package.json (hunk 16 above)
Comment (petebacondarwin) [reply package.json:39]: Yeah.
One problem with `npm` workspaces is that it is hard to delegate from a workspace up to the root when you run a script.

I was thinking of adding copies of these scripts to the `wrangler` workspace that worked, but then I was concerned that we would need to remember to keep everything up to date... 🤔

Encouraging all scripts to be run from the root made things simpler, yet more complicated...

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -1,12 +1,205 @@
...
+
+## Getting started
+
+### Set up your environment
+
+Wrangler2 is built and run on the Node.js JavaScript runtime.
+
+- Install the latest LTS version of [Node.js](https://nodejs.dev/) - we recommend using a Node version manager like [nvm](https://github.com/nvm-sh/nvm).
+- Install a code editor - we recommend using [VS Code](https://code.visualstudio.com/).
+  - Install the [recommended VS Code extensions](https://code.visualstudio.com/docs/editor/extension-marketplace#:~:text=install%20the%20recommended%20extensions) for the project.
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply CONTRIBUTING.md:15]: It currently suggests ESLint & Prettier
https://github.com/cloudflare/wrangler2/blob/main/.vscode/extensions.json#L2

File: CONTRIBUTING.md (hunk 15 above)
Comment (JacobMGEvans) [reply CONTRIBUTING.md:205]: Same 😄

File: CONTRIBUTING.md (hunk 13 above)
Comment (JacobMGEvans) [reply CONTRIBUTING.md:123]: We are also running these in CI, which removes/mitigates them as overhead from OSS contributors

File: CONTRIBUTING.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply CONTRIBUTING.md:15]: Good call!

File: package.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply package.json:39]: 👍

**Pull request merged**